### PR TITLE
Refactoring of useSetupTradeState

### LINKED ISSUE #2308: URL parameters are reset when follow a certain link
Reported in https://cowservices.slack.com/archives/C036MD8ADMG/p1681399726074199

**repro steps**:

1. open incognito window
2. navigate to: https://swap.cow.fi/#/100/swap/USDC/XDAI

**expected**: "gnosis chain", swap between usdc to xdai
**current**: "ethereum", swap between usdc to "select a token"

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes #2308

`useSetupTradeState()` hook serves to sync the URL with the app state.
The previous implementation was too complex, so, the hook was completely refactored.
Now it contains 3 `useEffect` hooks:
1. Reacts only to URL changes and does one of three actions:
a) Just set a state from URL to jotai
b) Reset the state to default
c) "Special case" - read more bellow

2. Reacts only on chanId changes in the URL  and does:
a) Just request network change the connected wallet
b) Use chainId from "Special case" if it's enabled

3. Reacts only on chanId changes in the wallet provider and does:
a) Reset the state to default corresponding to the new chainId
b) Apply "Special case"

> Important!
> The hook only works in one direction:
> user input (URL changes or network changes in the wallet) -> navigation -> jotai stateupdate

  # To Test
  
## When wallet is connected

1. And chainId is changed in URL, then should:
- [ ] revert URL to the previous state
- [ ] change network in the wallet
- [ ] apply the reverted state to the URL after changing chainId in the wallet

2. When wallet is connected and tokens are changed in URL, then should:
- [ ] update tokens in the state

3. When a wallet changed network, then should:
- [ ] reset state to default

## When wallet is NOT connected

1. And all params in URL are valid, then should:
- [ ] apply params from URL to the state

2. And chainId in URL is not valid, then should:
- [ ] reset the state to default

3. And the URL contains the same tokens (USDC/usdc), then should:
- [ ] reset the state to default

4. And the URL does not contain any tokens, then should:
- [ ] reset the state to default

5. And only chainId changed, then should:
- [ ] reset the state to default

6. And network changed from UI, then should:
- [ ] reset the state to default

### CODE REVIEW

File: src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts [hunk 1]
```diff
@@ -1,195 +1,168 @@
...
-   * Unlock network switching in the provider when chainId was changed in the URL
-   */
-  useLayoutEffect(() => {
-    if (chainIdFromUrlWasChanged) {
-      setIsChainIdSet(false)
-    }
-  }, [chainIdFromUrlWasChanged])
+    // Triggering only on changes from URL
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [tradeStateFromUrl])
```
Comment (alfetopito) [inline src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts:110]: Does this work? (haven't tested it yet)
In the hook `currentChainId` and `rememberedUrlState` are relied upon but they are not part of the hook dependencies.

File: src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts [hunk 2]
```diff
@@ -1,195 +1,168 @@
...
+    })
 
-    return () => {
-      isSubscribed = false
-    }
-  }, [isChainIdSet, chainIdFromUrl, connector, currentChainId, updateStateAndNavigate])
+    console.debug('[TRADE STATE]', 'Set chainId to provider', { connector, urlChainId })
+    // Triggering only when chainId in URL is changes and when connector is changed
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [connector, urlChainId, rememberedUrlState])
```
Comment (alfetopito) [inline src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts:140]: Not depending on `providerChainId`?

**Review by alfetopito — COMMENTED** [review]
Code review, will check the behaviour next

**Comment by alfetopito** [comment]
Found one use case where it's a bit broken.

Got there using wallect connect and metamask.

1. Change in the wallet to an unsupported network
* URL changes to this network's id
2. Change in the wallet to a supported network
* URL remains the unsupported network's
image

**Comment by elena-zh** [comment]
Hey @shoom3301 , great.
Some issues: 
1. Not connected: the URL is not reset to the default state for these cases
```
And chainId in URL is not valid, then should:
 reset the state to default
And the URL contains the same tokens (USDC/usdc), then should:
 reset the state to default
And the URL does not contain any tokens, then should:
 reset the state to default
And only chainId changed, then should:
 reset the state to default
```
when I press on the 'enter' keyboard button once (I change something in the URL and press on the Enter). But it works when I press on the Enter the 2nd time. Is this OK?

default1
enter
reset to dafault

2. When I paste this link https://swap-dev-git-fix-2308-cowswap.vercel.app/#/100/swap/ being not connected, I'm navigated to the https://swap-dev-git-fix-2308-cowswap.vercel.app/#/1/swap/WETH . Why can't we navigate a user to the Gnosis chain default state?
The same with https://swap-dev-git-fix-2308-cowswap.vercel.app/#/100/swap/WETH/xdai : The network is not picked, tokens do not appear in the Swap fields. 

3. Connected state: when change a network indicator, and cancel the request to change the network in the wallet. the network indicator in the URL is not changed back:
 
not returned back

4. Connected state: 
 - I'm connected to Gnosis chain
 - I paste https://swap-dev-git-fix-2308-cowswap.vercel.app/#/1/swap/eth/usdc link
 - I don't get the request to change the network to Mainnet and navigated to the GC default state


Thanks

**Comment by shoom3301** [comment]
@elena-zh @alfetopito thank you!
Fixed all

File: src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts [hunk 3]
```diff
@@ -1,195 +1,168 @@
...
+   * On URL parameter changes
+   *
+   * 1. SPECIAL CASE, when URL was changed while wallet is connected (read about it bellow)
+   * 2. When chainId in URL is invalid, then redirect to the default chainId
+   * 3. When URL contains the same token symbols (USDC/USDC), then redirect to the default state
+   * 4. When URL doesn't contain both tokens, then redirect to the default state
+   * 5. When only chainId was changed in URL, then redirect to the default state
+   * 6. Otherwise, fill the trade state by data from URL
+   *
+   * *** SPECIAL CASE ***
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts]: Thanks! Renamed

File: src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts (hunk 1 above)
Comment (shoom3301) [reply src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts:112]: Yes, it works

File: src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts (hunk 2 above)
Comment (shoom3301) [reply src/cow-react/modules/trade/hooks/setupTradeState/useSetupTradeState.ts:142]: Yes, it depends on only connector, urlChainId, rememberedUrlState

**Review by elena-zh — APPROVED** [review]
Great job!

**Review by alfetopito — APPROVED** [review]
Working great!

One final nitpick (which is the same behaviour as in prod right now) that does not have to be addressed in this PR

When I select a different network in the UI, and reject it, the error remains until I select another network and accept the change in the wallet.

If I change the networks in the wallet the error remains

https://user-images.githubusercontent.com/43217/233051668-d538d4c7-6ee3-46c3-83ca-c0dd9bbff90c.mov

Anyway, edge case, does not impact the app, only the error is visible.
Can be put in the backlog with low priority

**Pull request merged**